internal/cloud/awscloud/secure-instance.go

```diff
@@ -0,0 +1,378 @@
+package awscloud
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/sirupsen/logrus"
+)
+
+type SecureInstance struct {
+	FleetID  string
+	SGID     string
+	LTID     string
+	Instance *ec2.Instance
+}
+
+func (a *AWS) RunSecureInstance(iamProfile string) (*SecureInstance, error) {
+	identity, err := a.ec2md.GetInstanceIdentityDocument()
+	if err != nil {
+		logrus.Errorf("Error getting the identity document, %s", err)
+		return nil, err
+	}
+
+	descrInstancesOutput, err := a.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(identity.InstanceID),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(descrInstancesOutput.Reservations) != 1 || len(descrInstancesOutput.Reservations[0].Instances) != 1 {
+		return nil, fmt.Errorf("Expected exactly one instance")
+	}
+	vpcID := *descrInstancesOutput.Reservations[0].Instances[0].VpcId
+	imageID := *descrInstancesOutput.Reservations[0].Instances[0].ImageId
+
+	var secureInstance *SecureInstance
+	defer func() {
+		if secureInstance != nil && secureInstance.Instance == nil {
+			if secureInstance.FleetID != "" {
+				delFlOutput, err := a.ec2.DeleteFleets(&ec2.DeleteFleetsInput{
+					FleetIds: []*string{
+						aws.String(secureInstance.FleetID),
+					},
+					TerminateInstances: aws.Bool(true),
+				})
+				if err != nil {
+					logrus.Errorf("Deleting fleet unsuccessful: %v", err)
+				}
+				if len(delFlOutput.UnsuccessfulFleetDeletions) != 0 || len(delFlOutput.SuccessfulFleetDeletions) != 1 {
+					logrus.Error("Deleting fleet unsuccessful")
+				}
+			}
+			if secureInstance.LTID != "" {
+				_, err := a.ec2.DeleteLaunchTemplate(&ec2.DeleteLaunchTemplateInput{
+					LaunchTemplateId: aws.String(secureInstance.LTID),
+				})
+				if err != nil {
+					logrus.Errorf("Error deleting  launch template %s: %v", secureInstance.LTID, err)
+				}
+			}
+			if secureInstance.SGID != "" {
+				// security group was created but instance wasn't
+				_, err := a.ec2.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+					GroupId: aws.String(secureInstance.SGID),
+				})
+				if err != nil {
+					logrus.Errorf("Error deleting security group %s: %v", secureInstance.SGID, err)
+				}
+			}
+		}
+	}()
+
+	sgID, err := a.createOrReplaceSG(identity.InstanceID, identity.PrivateIP, vpcID)
```

Comment:
I had to dig into this a little before I understood that sdID could be set, and need cleaning up, even when there is an error. Maybe a comment about that is needed ?
Also, it seems (from the outside of the function) fragile to depend on an if block to create the SecureInstance, is there any reason not to just make an empty one and fill it in?

Reply:
I tried to make it a bit clearer, and construct a default SI now.

Reply:
I tried to restructure the cleanup a bit, and added a small comment on top of the fn.